cli/src/action/login/oauth/github.rs

```diff
@@ -0,0 +1,238 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use oauth2::basic::BasicClient;
+use oauth2::{
+    AccessToken, AuthUrl, AuthorizationCode, ClientId, ClientSecret, CsrfToken, PkceCodeChallenge,
+    RedirectUrl, Scope, TokenResponse, TokenUrl,
+};
+use reqwest::blocking::Client;
+
+use super::callback;
+use super::{OAuth2Provider, OAuth2ProviderError, ProviderType, UserTokens};
+
+pub struct GithubProvider {
+    client_id: Option<&'static str>,
+    client_secret: Option<&'static str>,
+}
+
+impl OAuth2Provider for GithubProvider {
+    fn get_tokens(&self) -> Result<UserTokens, OAuth2ProviderError> {
+        let client_id = self.client_id.ok_or_else(|| {
+            OAuth2ProviderError("Github client id was not configured for this binary".into())
+        })?;
+        let client_secret = self.client_secret.ok_or_else(|| {
+            OAuth2ProviderError("Github client secret was not configured for this binary".into())
+        })?;
+
+        let callback = callback::OAuth2Callback::new().expect("Unable to start callback");
+
+        let auth_url = AuthUrl::new("https://github.com/login/oauth/authorize".to_string())
+            .map_err(|err| OAuth2ProviderError(err.to_string()))?;
+        let token_url = TokenUrl::new("https://github.com/login/oauth/access_token".to_string())
+            .map_err(|err| OAuth2ProviderError(err.to_string()))?;
+        let redirect_url = RedirectUrl::new(format!("http://{}/auth_response", callback.address()))
+            .map_err(|err| OAuth2ProviderError(err.to_string()))?;
+
+        let client = BasicClient::new(
+            ClientId::new(client_id.to_string()),
+            Some(ClientSecret::new(client_secret.to_string())),
+            auth_url,
+            Some(token_url),
+        )
+        .set_redirect_url(redirect_url);
+
+        // Github supports PKCE challenge.
+        let (pkce_challenge, pkce_verifier) = PkceCodeChallenge::new_random_sha256();
+
+        // Generate the full authorization URL.
+        let (auth_url, _csrf_token) = client
+            .authorize_url(CsrfToken::new_random)
+            // Set the desired scopes.
+            .add_scope(Scope::new("user:email".to_string()))
+            // Set the PKCE code challenge.
+            .set_pkce_challenge(pkce_challenge)
+            .url();
+
+        info!("Opening {}...", auth_url);
+        webbrowser::open(&auth_url.to_string()).map_err(|err| {
+            debug!("Unable to open authorization URL: {}", err);
```

Comment:
Is this debug necessary with the error being returned? Seems like it would result in a double log.

Reply:
The debug flag is there to give the details of the source error.  If we are trying to move away from creating the poor-man's stack trace, by appending strings to the messages, then we still need a way to get the details out.  Presumable the error returned would be logged as error, where this one would be logged if debug details are turned on (i.e. someone is debugging)